File: email-aggregator-go/src/events/bus.go
package events

import (
	"context"
	"time"
)

// EventBus 消息总线接口（at-least-once 语义，消费端自行幂等）
type EventBus interface {
	Publish(ctx context.Context, topic, key string, payload []byte) error
	Subscribe(ctx context.Context, topic, group string, handler func(ctx context.Context, env EventEnvelope) error) error
	Close() error
}

// InMemoryBus 进程内实现（PoC / 单测 / 本地无外部依赖跑通）
type InMemoryBus struct {
	subs map[string][]func(ctx context.Context, env EventEnvelope) error
}

// NewInMemoryBus 构造
func NewInMemoryBus() *InMemoryBus {
	return &InMemoryBus{subs: map[string][]func(ctx context.Context, env EventEnvelope) error{}}
}

// Publish 同步 fan-out 给订阅者（演示用；生产由 Kafka 分区保证顺序与重放）
func (b *InMemoryBus) Publish(ctx context.Context, topic, key string, payload []byte) error {
	env := EventEnvelope{Topic: topic, Key: key, TS: time.Now().UnixMilli(), Type: TypeOfTopic(topic), Payload: payload}
	for _, h := range b.subs[topic] {
		if err := h(ctx, env); err != nil {
			return err
		}
	}
	return nil
}

// Subscribe 注册处理器（同主题可多 group 消费）
func (b *InMemoryBus) Subscribe(_ context.Context, topic, _ string, handler func(ctx context.Context, env EventEnvelope) error) error {
	b.subs[topic] = append(b.subs[topic], handler)
	return nil
}

// Close 释放
func (b *InMemoryBus) Close() error { return nil }

// 真实 Kafka 适配器见 integration/integration.go 中的 KafkaAdapter（//go:build integration）。
// 接口与 InMemoryBus 完全一致，替换时无需改动编排/采集逻辑。

File: email-aggregator/src/events/bus.ts
/**
 * 事件总线。PoC 以 InMemoryEventBus 实现（at-least-once + 幂等去重 + DLQ）。
 * KafkaAdapter 为真实集成 stub：接口一致，落地时替换为 kafkajs 客户端。
 */

import type { AppEvent, Topic } from "./contracts.ts";

export type Handler = (e: AppEvent) => Promise<void> | void;

export interface EventBus {
  publish(event: AppEvent): Promise<void>;
  subscribe(topic: Topic, handler: Handler): void;
  /** 启动消费（InMemory 立即开始派发已注册订阅） */
  start(): Promise<void>;
  stop(): Promise<void>;
}

export class InMemoryEventBus implements EventBus {
  private handlers = new Map<Topic, Handler[]>();
  private seen = new Set<string>(); // idempotencyKey 去重
  private dlq: AppEvent[] = [];
  private running = false;

  subscribe(topic: Topic, handler: Handler): void {
    const list = this.handlers.get(topic) ?? [];
    list.push(handler);
    this.handlers.set(topic, list);
  }

  async publish(event: AppEvent): Promise<void> {
    // 至少一次语义：先记录、再派发；幂等由消费端 seen 集合保证
    if (!this.running) return; // demo 显式 start 后开始派发
    const topic = event.topic;
    const list = this.handlers.get(topic) ?? [];
    for (const h of list) {
      try {
        // 幂等去重
        if (this.seen.has(event.idempotencyKey)) continue;
        this.seen.add(event.idempotencyKey);
        await h(event);
      } catch (err) {
        // 失败进 DLQ，等待人工 / 重试（PoC 仅记录）
        this.dlq.push(event);
        console.error(`[bus] handler failed on ${topic}:`, (err as Error).message);
      }
    }
  }

  async start(): Promise<void> {
    this.running = true;
  }

  async stop(): Promise<void> {
    this.running = false;
  }

  getDlq(): AppEvent[] {
    return this.dlq;
  }
}

/**
 * @deprecated STUB — 未实现。仅作接口文档保留。
 * 真实 Kafka 适配器落地时改为：
 *   import { Kafka } from "kafkajs";
 *   producer.send({ topic, messages: [{ key: partitionKey, value: JSON.stringify(event) }] })
 * 消费端以 consumer-group 实现再均衡（同 accountId 固定分区 → 单账户有序）。
 * Go 版真实实现见 integration/integration.go 中的 KafkaAdapter。
 */
export class KafkaAdapter implements EventBus {
  private brokers: string[];
  constructor(brokers: string[]) {
    this.brokers = brokers;
    console.warn(
      `[KafkaAdapter] stub: 未连接 ${brokers.join(",")}，请接入 kafkajs 实现 publish/subscribe`,
    );
  }
  async publish(_event: AppEvent): Promise<void> {
    throw new Error("KafkaAdapter.publish not implemented");
  }
  subscribe(_topic: Topic, _handler: Handler): void {
    /* noop stub */
  }
  async start(): Promise<void> {}
  async stop(): Promise<void> {}
}
